go-client/ring.go

```diff
@@ -0,0 +1,218 @@
+}
+
+type HashAlgo string
+
+var (
+	Rendezvous HashAlgo = "rendezvous"
+	Jump       HashAlgo = "jump"
+	Maglev     HashAlgo = "maglev"
+)
+
+type ConsistentHash interface {
+	Lookup(key string) string
+}
+
+type Ring struct {
+	opts                RingOptions
+	consistentHash      ConsistentHash
+	resetConsistentHash func(addrs []string) ConsistentHash
+	allShards           map[string]*shardClient
+	mu                  sync.RWMutex
+	commandAdaptor
+}
+
+type shardClient struct {
+	client    *Client
+	downCount int32
+}
+
+func NewDefaultRing(addr []string) (*Ring, error) {
+	defaultOpts := RingOptions{
+		HealthCheckFrequency: 30 * time.Second,
+		HashAlgorithm:        Rendezvous,
+		ClientOptions: RingClientOpt{
+			PoolTimeout: time.Second, // max time to wait to get new connection from pool
+			PoolSize:    20,          // max number of connection can get from the pool
+			MinIdleConn: 5,
+			Dialer: func(ctx context.Context, addr string) (net.Conn, error) { //Must define dialer func
+				conn, err := net.Dial("tcp", addr)
+				if err != nil {
+					return nil, err
+				}
+				return conn, err
+			},
+			IdleTimeout:        time.Minute * 5,  // if connection lives longer than 5 minutes, it is removable
+			MaxConnAge:         time.Minute * 10, // all connections cannot live longer than this
+			IdleCheckFrequency: time.Minute * 5,  // reap staled connections after 5 minutes
+		},
+	}
+	defaultOpts.Addresses = addr
+	return NewRing(defaultOpts)
+}
+
+// NewRing return a Ring behaves like a single client, with consistent hashing
+func NewRing(opts RingOptions) (*Ring, error) {
+	r := &Ring{
+		opts: opts,
+		mu:   sync.RWMutex{},
+	}
+	switch opts.HashAlgorithm {
+	case Rendezvous:
+		r.resetConsistentHash = func(addrs []string) ConsistentHash {
+			return newRendezvous(addrs)
+		}
+	default:
+		return nil, fmt.Errorf("unsupported hash algorithm: %s", opts.HashAlgorithm)
+	}
+	allShards := make(map[string]*shardClient)
+	clientOpt := opts.ClientOptions
+	for _, addr := range opts.Addresses {
+		clientOpt := ClientOptions{
+			Pool: pool.Options{
+				Address:            addr,
+				PoolTimeout:        clientOpt.PoolTimeout,
+				PoolSize:           clientOpt.PoolSize,
+				MinIdleConn:        clientOpt.MinIdleConn,
+				Dialer:             clientOpt.Dialer,
+				IdleTimeout:        clientOpt.IdleTimeout,
+				MaxConnAge:         clientOpt.MaxConnAge,
+				IdleCheckFrequency: clientOpt.IdleCheckFrequency,
+			},
+		}
+		client, err := NewClient(clientOpt)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create client for address %s: %s", addr, err)
+		}
+		allShards[addr] = &shardClient{
+			downCount: 0,
+			client:    client,
+		}
+	}
+	r.allShards = allShards
+	r.updateAliveShards()
+	go r.healthCheckByInterval(opts.HealthCheckFrequency)
+	r.commandAdaptor = r.cmdWithConsistentHash
+	return r, nil
+}
+
+func (r *Ring) Close() error {
+	return nil
+	//TODO
+}
+
+func (r *Ring) cmdWithConsistentHash(cmd Cmd) error {
+	r.mu.RLock()
```

Comment:
Should move `r.mu.RUnlock()` below `RLock()`, and should you `defer r.mu.RUnlock()`

Reply:
This is used to release the lock early, using defer means the lock is released only after the response from keva is returned.

Reply:
Ah oh, got it